### ISSUE TITLE
Standardize Surface-Measurements API (feature request)

### ISSUE THREAD

**Issue by haesleinhuepf** [morphometrics/morphometrics#45]
Hi @kevinyamauchi , hi @jo-mueller,

I hope you're doing well. As a follow up to our hackathon, I would like to start a discussion on standardizing an API for surface measurements. So far, `morphometrics` standardizes the measurement-API so that all functions with the signature `def my_measurement(image:ndarray, labels:ndarray, feature1:bool, feature2:bool, ...)` can be registered. Example:
https://github.com/morphometrics/morphometrics/blob/f39eac8b40eb491931a1fa08d1227c86bcf0522e/src/morphometrics/measure/label.py#L15-L22

I'm wondering if we can extend the very same `register_measurement_set()`-strategy so that we can also handle surface-based measurements, or if we can come up with a compatible counterpart for surface-based measurements. Also point-cloud-based measurements could be handled similarly.

I guess we should also standardize what the default surface- and point-cloud datatypes are. There is [at least one function](https://github.com/morphometrics/morphometrics/blob/f39eac8b40eb491931a1fa08d1227c86bcf0522e/src/morphometrics/measure/surface.py#L10) in `morphometrics` where the surface comes as `trimesh.Trimesh`, In napari-stress we use napari's `VectorsData` and `PointsData` to do measurements ([exmple](https://github.com/BiAPoL/napari-stress/blob/main/src/napari_stress/_measurements/curvature.py#L26)), and in napari-process-points-and-surfaces, we use `SurfaceData` ([example](https://github.com/haesleinhuepf/napari-process-points-and-surfaces/blob/b1eee0567e9b879e8c873d7fc3c0a01752df039c/napari_process_points_and_surfaces/_quantification.py#L79)).

I would love to standardize these things so that we can easily join many measurements in one table and apply them from a joint user-interface.

Let me know what you think!

Best,
Robert

**Comment by kevinyamauchi** [comment]
Hey @haesleinhuepf ! I agree that the measurement API should be expanded to include measurements on surfaces. I am not sure the best strategy here. My first thoughts are below. I am curious to hear your thoughts:

- we should break out a small package for just the measurement engine. That way people don't have to depend on morphometrics, which has an openCL dependency. The measurement engine package can be super lightweight (pure python?).
- I think we should have different registries for different types of measurements. In this case, "type" means function signatures. Some examples:   
    - `label_measurements`  which take a `LabelImage` and optional `IntensityImage` and return a `LabelTable` (this is what we currently have)
    - `surface_vertex_measurements` which takes a `Mesh` and optional `IntensityImage` and returns a `VertexTable` (each row has  measurements for a given vertex).
    - `surface_face_measurements`, which takes which takes a `Mesh` and optional `IntensityImage` and returns a `FaceTable` (each row has  measurements for a given face)

I am also into standardizing the data model for meshes. I currently use trimesh because it's relatively easy to install and can do most standard operations. The downside is that it can be slow. I am definitely open to other data models. See [this related issue](https://github.com/morphometrics/morphometrics/issues/31) on the data model for surfaces/meshes and I can quickly summarize my thoughts here:
- I think we want something that allows us to access vertices, faces, and can store measurement values.
- I think we want to be able to store annotations/measurements on both the faces and the vertices
- I am open to wrapping the mesh object (e.g., vedo, pyvista) from an existing library with some convenience methods if that's the easiest approach.
- for compatibility, it would be nice if there was a `to_napari_surface_data()` method.

**Comment by haesleinhuepf** [comment]
> * we should break out a small package for just the measurement engine. That way people don't have to depend on morphometrics, which has an openCL dependency. The measurement engine package can be super lightweight (pure python?).

Fully agreed, a minimal package with pure python and hardly any dependencies would be good.

> I am definitely open to other data models.

Is a pure numpy-derived data type (similar to `napari.types.SurfaceData` and `napari.types.PointsData`) an option? These are easy to convert to other libraries (open3d, vedo, ...) and have no dependencies. It's just numpy.

**Comment by kevinyamauchi** [comment]
> Is a pure numpy-derived data type (similar to napari.types.SurfaceData and napari.types.PointsData) an option? These are easy to convert to other libraries (open3d, vedo, ...) and have no dependencies. It's just numpy.

Yes, I think this makes sense!

**Comment by kevinyamauchi** [comment]
> Fully agreed, a minimal package with pure python and hardly any dependencies would be good.

I've made an initial prototype which should replicate the current functionality in `morphometrics`: https://pypi.org/project/morphometrics-engine/

**Comment by haesleinhuepf** [comment]
In [napari-stress](https://github.com/BiAPoL/napari-stress) you're using a `List[SurfaceData]` for this, right @jo-mueller ?

**Comment by jo-mueller** [comment]
Yes. Which is nice but then again also unsatisfactory, I'd prefer to access `SurfaceData` with time as a numpy array as 
 `my_surface_data[timepoint, points, faces]` - which would point towards designing our own Surface wrapper class as suggested by @kevinyamauchi above. For this, we could simply overwrite `__get__()` to have the indexing behavior we'd want.

As for the measurements - I'm not sure, but could it also be desirable to assign measurements to edges?

**Comment by haesleinhuepf** [comment]
I'm afraid that
```
my_surface_data[timepoint, points, faces]
```
is incompatible with numpy. It would only work if the number of points is the same in all timepoints. If anyone can demonstrate that I'm wrong, e.g. with some example code, I'd be happy to learn something 🌞

**Comment by jo-mueller** [comment]
I was think about something like this:

```Python
class SurfaceWrapper():
    def __init__(self, surface_data: SurfaceData):
        self.Converter = TimelapseConverter()
        self.data = Converter.data_to_list_of_data(surface_data, SurfaceData)
    def __getitem__(self, i):
        substack = self.Converter.list_of_data_to_data(self.data[i], layertype=SurfaceData)
        return substack
    def __len__(self):
        return len(self.data)
```

whereas the `TimelapseConverter` comes from napari-stress and allows list-based SurfaceData slicing, so that could be solved like that. I guess the next problem would be that the number of vertices and faces doesn't have to be equal so this would have to be a tuple.

**Comment by haesleinhuepf** [comment]
I'm wondering if `List[Tuple[vertices,faces,values]]` with vertices, faces and values being numpy [compatible] arrays, could be a good defined standard data type for our purpose here. We can of course convert from/to everything. I'm proposing a basic python/numpy-derived type in order to avoid https://xkcd.com/927/

**Comment by jo-mueller** [comment]
Point taken 😆
